### PR TITLE
Supporting multiple functions (Part II & III): pure functions and function calls

### LINKED ISSUE #56: Add multiple functions to RVSDG optimizer

### PR DESCRIPTION
**Pull request by yihozhang**
This PR makes three changes:

* To better optimize pure functions and function calls, this PR adds a stateful flag to each function call and each function declaration. As a result, each function call may produce one to two results depending on the function type:
  * StateAndValue
  * StateOnly
  * ValueOnly
* Earlier conversion PRs (#44) do not support translating function calls back to CFG. This PR fixed it.
* Finally, this PR completes the two incomplete tests in the `failing` folder and moves one out of `failing` since it is passing right now.

Fixes #56

### CODE REVIEW

**Comment by yihozhang** [comment]
Don't merge this until #77 is merged

**Comment by yihozhang** [comment]
Marking it as draft as I'll be refactoring it to reflect @oflatt's proposal of un-specializing state edges.

**Comment by yihozhang** [comment]
@oflatt I've refactored the code according to what you asked for during our discussion. Something to note here:

* You mentioned during the discussion that we don't need to track how many arguments a function call can return, since we can always look up the callee definition. However, unfortunately, currently, the to_cfg code does not have access to function definitions, so we have to keep n_outs in our egglog encoding, which is actually consistent with the Rust definition of `BasicExpr::Call`.
* A function can return multiple values just like Gamma and Theta. However, only one of the return values is a Bril value (that is, a non-state value). Again, because we don't have access to function definitions during rvsdg-to-cfg, we don't know where this Bril value is (i.e. we don't know which `Proj` produces the actual return value). So for now we have to assume that the return value is the first element of return values for code generation to work.

I think one way to solve the above two issues is to keep the type signature of the callee in `BasicExpr::Call`, which I don't think is a very unusual thing to do. What do you think?

File: src/cfg/tests.rs [hunk 1]
```diff
@@ -71,26 +71,27 @@ cfg_test!(
...
+//         "for.cond" = (Cond { arg: "for_cond_0".into(), val: true.into() }) => "for.body",
+//         "for.cond" = (Cond { arg: "for_cond_0".into(), val: false.into() }) => "next.ret.1",
+//         "for.body" = (Cond { arg: "is_valid".into(), val: true.into() }) => "rec.func",
+//         "for.body" = (Cond { arg: "is_valid".into(), val: false.into() }) => "next.loop",
+//         "rec.func" = (Jmp) => "next.loop",
+//         "next.loop" = (Jmp) => "for.cond",
+//         "next.ret" = (Jmp) => EXIT,
+//         "next.ret.1" = (Jmp) => EXIT,
+//     ]
+// );
```
[the remark this answers is not in the document]
Comment (yihozhang) [reply src/cfg/tests.rs]: @Alex-Fischman This is fixed now.

**Review by Alex-Fischman — COMMENTED** [review]
LGTM

**Review by oflatt — APPROVED** [review]
Amazing work @yihozhang! Thanks for changing based on feedback

**Pull request merged**